src/uct/ib/base/ib_md.c

```diff
@@ -74,6 +73,275 @@ static ucs_status_t uct_ib_md_query(uct_md_h uct_md, uct_md_attr_t *md_attr)
+    } else {
+        qp_attr.ah_attr.is_global    = 0;
+    }
+#endif
+
+    qp_attr.rq_psn                   = 0;
+    qp_attr.path_mtu                 = IBV_MTU_512;
+    qp_attr.min_rnr_timer            = 7;
+    qp_attr.max_dest_rd_atomic       = 1;
+    ret = ibv_modify_qp(md->umr_qp, &qp_attr,
+                        IBV_QP_STATE | IBV_QP_AV | IBV_QP_PATH_MTU | IBV_QP_DEST_QPN |
+                        IBV_QP_RQ_PSN | IBV_QP_MAX_DEST_RD_ATOMIC | IBV_QP_MIN_RNR_TIMER);
+    if (ret) {
+        ucs_error("Failed to modify UMR QP to RTR: %m");
+        goto err_destroy_qp;
+    }
+
+    /* Modify to RTS */
+    qp_attr.qp_state                 = IBV_QPS_RTS;
+    qp_attr.sq_psn                   = 0;
+    qp_attr.timeout                  = 7;
+    qp_attr.rnr_retry                = 7;
+    qp_attr.retry_cnt                = 7;
+    qp_attr.max_rd_atomic            = 1;
+    ret = ibv_modify_qp(md->umr_qp, &qp_attr,
+                        IBV_QP_STATE | IBV_QP_TIMEOUT | IBV_QP_RETRY_CNT |
+                        IBV_QP_RETRY_CNT | IBV_QP_RNR_RETRY | IBV_QP_SQ_PSN |
+                        IBV_QP_MAX_QP_RD_ATOMIC);
+    if (ret) {
+        ucs_error("Failed to modify UMR QP to RTS: %m");
+        goto err_destroy_qp;
+    }
+    return UCS_OK;
+
+err_destroy_qp:
+    ibv_destroy_qp(md->umr_qp);
+err_destroy_cq:
+    ibv_destroy_cq(md->umr_cq);
+err:
+    return UCS_ERR_IO_ERROR;
+#else
+    md->umr_qp = 0;
+    md->umr_cq = 0;
+    return UCS_ERR_NOT_IMPLEMENTED;
+#endif
+}
+
+static void uct_ib_md_umr_qp_destroy(uct_ib_md_t *md) 
+{
+#if HAVE_EXP_UMR
+    if (md->umr_qp) {
+        ibv_destroy_qp(md->umr_qp);
+    }
+    if (md->umr_cq) {
+        ibv_destroy_cq(md->umr_cq);
+    }
+#endif
+}
+
+uint8_t  uct_ib_md_umr_id(uct_ib_md_t *md)
+{
+#if HAVE_EXP_UMR
+    uint8_t umr_id;
+
+    if (md->umr_qp == NULL || md->umr_cq == NULL) {
+        return 0;
+    }
+    /* Generate umr id. We want umrs for same virtual adresses to have different
+     * ids across proceses.
+     * 
+     * Usually parallel processes running on the same node as part of a single
+     * job will have consequitive pids. For example mpi ranks, slurm spawned tasks...
+     *
+     * Theoretically it is possible to have different ids for 512 processes (4096/8)
+     * but we only use 256 so that we can fit into a byte
+     */
+    umr_id = getpid() % 256;
+    return umr_id;
+#else
+    return 0;
+#endif
+}
+
+static UCS_F_MAYBE_UNUSED
+struct ibv_mr *uct_ib_md_create_umr(uct_ib_md_t *md, struct ibv_mr *mr)
+{
+#if HAVE_EXP_UMR
+    struct ibv_exp_mem_region mem_reg;
+    struct ibv_exp_send_wr wr, *bad_wr;
+    struct ibv_exp_create_mr_in mrin;
+    struct ibv_mr *umr;
+    struct ibv_wc wc;
+    int ret;
+    ucs_status_t status;
+    unsigned count;
+    size_t offset;
+
+    if (md->umr_qp == NULL || md->umr_cq == NULL) {
+        return NULL;
+    }
+    /* note: all errors are fatal because we have a single umr_offset per process. 
```

Comment:
the function can return status code, in case UMR fails (and umr_qp != NULL) memory registration should fail, does not have to be fatal. also note umr_offset is per device (md) and not per process since one process may be using 2 devices - one which supports umr and another which doesn't